view/src/storage.rs

```diff
@@ -666,6 +697,97 @@ impl Storage {
         Ok(())
     }
 
+    pub async fn give_advice(&self, note: Note) -> anyhow::Result<()> {
+        let mut tx = self.pool.begin().await?;
+
+        let note_commitment = note.commit().0.to_bytes().to_vec();
+        let address = note.address().to_vec();
+        let amount = u64::from(note.amount()) as i64;
+        let asset_id = note.asset_id().to_bytes().to_vec();
+        let blinding_factor = note.note_blinding().to_bytes().to_vec();
+
+        sqlx::query!(
+            "INSERT INTO notes
+                    (
+                        note_commitment,
+                        address,
+                        amount,
+                        asset_id,
+                        blinding_factor
+                    )
+                    VALUES
+                    (?, ?, ?, ?, ?)",
+            note_commitment,
+            address,
+            amount,
+            asset_id,
+            blinding_factor,
+        )
+        .execute(&mut tx)
+        .await?;
+
+        tx.commit().await?;
+
+        Ok(())
+    }
+
+    /// Return advice about note contents for use in scanning.
+    ///
+    /// Given a list of note commitments, this method checks whether any of them
+    /// correspond to notes that have been recorded in the database but not yet
+    /// observed during scanning.
+    pub async fn scan_advice(
+        &self,
+        note_commitments: Vec<note::Commitment>,
+    ) -> anyhow::Result<BTreeMap<note::Commitment, Note>> {
+        if note_commitments.is_empty() {
+            return Ok(BTreeMap::new());
+        }
+
+        let rows = sqlx::query(
+            format!(
+                "SELECT notes.address,
+                        notes.amount,
+                        notes.asset_id,
+                        notes.blinding_factor
+                FROM notes
+                LEFT OUTER JOIN spendable_notes ON notes.note_commitment = spendable_notes.note_commitment
+                WHERE (spendable_notes.note_commitment IS NULL) AND (notes.note_commitment IN ({}))",
+                note_commitments
+                    .iter()
+                    .map(|cm| format!("x'{}'", hex::encode(cm.0.to_bytes())))
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            )
```

Comment:
Followup: document this SQL query and its rationale.

We want to only give advice about notes that are known (hence in `notes`) but not already recorded as spendable, in which case we shouldn't see them again in scanning anyways, and if we do, something bad is probably happening.  I didn't think it through carefully, but I was worried that since the `spendable_notes` table records our own scan results, we might open up the possibility of leaking information about our scan results to an adversary who sends us crafted `CompactBlock`, e.g., sending a compact block with a `StatePayload::RolledUp(target_commitment)` and then somehow observing that we'd detected it using our scan results as advice.